File: src/main/java/edu/missouristate/mars/mips/instructions/syscalls/SyscallPrintDouble.java
package edu.missouristate.mars.mips.instructions.syscalls;

import edu.missouristate.mars.util.*;
import edu.missouristate.mars.mips.hardware.*;
import edu.missouristate.mars.*;

/**
 * Service to display double whose bits are stored in $f12 & $f13 onto the console.
 * $f13 contains high order word of the double.
 */

public class SyscallPrintDouble extends AbstractSyscall {
    /**
     * Build an instance of the Print Double syscall.  Default service number
     * is 3 and name is "PrintDouble".
     */
    public SyscallPrintDouble() {
        super(3, "PrintDouble");
    }

    /**
     * Performs syscall function to print double whose bits are stored in $f12 & $f13.
     */
    public void simulate(ProgramStatement statement) throws ProcessingException {
        // Note: Higher numbered reg contains high order word so concat 13-12.
        SystemIO.printString(Double.valueOf(Double.longBitsToDouble(
                Binary.twoIntsToLong(Coprocessor1.getValue(13), Coprocessor1.getValue(12))
        )).toString());
    }
}

File: src/main/java/edu/missouristate/mars/mips/instructions/syscalls/SyscallPrintDouble.kt
/*
 * Copyright (c) 2003-2024, Pete Sanderson and Kenneth Vollmar
 * Copyright (c) 2024-present, Nicholas Hubbard
 *
 * Originally developed by Pete Sanderson (psanderson@otterbein.edu) and Kenneth Vollmar (kenvollmar@missouristate.edu)
 * Maintained by Nicholas Hubbard (nhubbard@users.noreply.github.com)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
 * documentation files (the "Software"), to deal in the Software without restriction, including without limitation the
 * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to
 * permit persons to whom the Software is furnished to do so, subject to the following conditions:
 *
 * 1. The above copyright notice and this permission notice shall be included in all copies or substantial portions of
 *    the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
 * WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
 * OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

package edu.missouristate.mars.mips.instructions.syscalls

import edu.missouristate.mars.ProcessingException
import edu.missouristate.mars.ProgramStatement
import edu.missouristate.mars.bitsToDouble
import edu.missouristate.mars.mips.hardware.Coprocessor1.getValue
import edu.missouristate.mars.util.Binary.twoIntegersToLong
import edu.missouristate.mars.util.SystemIO

/**
 * Service to display double whose bits are stored in $f12 & $f13 onto the console.
 * $f13 contains high-order word of the double.
 */
class SyscallPrintDouble : AbstractSyscall(3, "PrintDouble") {
    /**
     * Performs syscall function to print double whose bits are stored in $f12 & $f13.
     */
    @Throws(ProcessingException::class)
    override fun simulate(statement: ProgramStatement) {
        // Note: Higher-numbered register contains the high-order word, so concat 13-12.
        SystemIO.printString(
            twoIntegersToLong(getValue(13), getValue(12)).bitsToDouble().toString()
        )
    }
}